tests/test_nfft.py

```diff
@@ -1,55 +1,454 @@
-import os.path
+import numpy as np
+import pytest
+
 from unittest import TestCase, skipUnless
 from unittest.case import SkipTest
 
-import numpy as np
+from aspire.nfft import backend_available
+from aspire.nfft import Plan
+from aspire.utils.misc import complex_type
 
-from aspire.nfft import Plan, backend_available
 
+import os.path
 DATA_DIR = os.path.join(os.path.dirname(__file__), 'saved_test_data')
 
 
 class SimTestCase(TestCase):
+
     def setUp(self):
-        pass
+        self.fourier_pts = np.array([
+            [ 0.88952655922411,  0.35922344760724, -0.17107966400962, -0.70138277562649],
+            [ 1.87089316522016,  1.99362869011803,  2.11636421501590,  2.23909973991377],
+            [-3.93035749861843, -3.36417300942290, -2.79798852022738, -2.23180403103185]
+        ])
+
+        self.vol = np.load(os.path.join(DATA_DIR, 'nfft_volume.npy'))
+
+        # Setup a 2D slice for testing 2d many
+        self.plane = self.vol[0]
+
+        self.recip_space = np.array([-0.05646675 + 1.503746j, 1.677600 + 0.6610926j, 0.9124417 - 0.7394574j, -0.9136836 - 0.5491410j])
+
+        self.recip_space_plane = np.array([-0.14933796+0.00324942j,
+                                           -0.29726508+0.07601329j,
+                                           -0.26276074+0.12634184j,
+                                           -0.09722212+0.12028122j],
+                                          dtype=np.complex128)
+
+        self.adjoint_vol = np.array([[[9.59495207e-01-1.06291322e+00j,  4.96001394e-01+1.60922425e+00j,
```

Comment:
Can we make the self.adjoint_vol smaller or in some way else? There are more than 200 lines.

Reply:
In spirit I agree totally, and I'd probably throw out most of the existing and new unit tests wholesale, but its not something I want to do in this PR. At this point with what we're queuing up to change (the storage ordering), I don't agree in practice we should address it in this PR. It's by far not the worst offender of this sort :)...

Im not really sure it is effective use of time to remark on the items that were already identified as being a top level issue. I'll copy point 1 above here and add some color so you understand I agree with you.
"""
The unit tests are totally gross, but we have coverage now where we did not before. Once all the code is internally consistent I think it would be easier to just do a more typical adjoint based test (the dot trick), and maybe convert to a test matrix style for that group of tests.
""" (notice I call my own code gross, I know this is not what I want)

By internally consistent, I mean regarding things like storage, etc.  If we do this improvement now, before that is reasonably consistent, we will just have to redo the unit tests, again, following those changes. Seems like a poor plan to me. We need to think a couple steps ahead about this stuff.

By matrix style, I mean a more expressive way to cover all the permutations of backend X dtype X dimension X manyness since most of it is redundant (the code here is very template-ish).

It is possible to do this now, but it will be manufacturing rework imo. I'm mainly worried with any major defects, install process, or user facing problems (besides the ordering which is waiting on everything outstanding to start making passes..).  Sometimes it is more efficient to make many small waves of changes quickly then a tsunami twice a year. I expect to his the identified points at the right time.

Reply:
Ok. Can you make an issue so that we have can add it to the board?

Reply:
Added #229. I'm not sure we need for the first release. The tests are totally serviceable besides aesthetics. When I came to this code it was essentially running a single test for one case. It is now running forward, backward X dtypes x many. I could have spent that time to make one pretty test that was useless. I stand by my decision.

Reply:
I agree. Just wanted to make sure we have an issue on it.

Reply:
I am not sure what the criteria for the unit test now. We used to try to do in this way. If data set is small like less tens of lines, we will include directly in the code. Otherwise we will put it in an additional file under the tests/saved_test_data if we can not find an easy way to make it small. In future we might separate the tests and tutorials to a separate repo.